File: src/main/problems/java/Nr206.java
package main.problems.java;

import main.util.java.ListNode;
import main.util.java.ListNodeBuilder;
import main.util.java.ListNodePrinter;

public class Nr206 {

    public static void main(String[] args) {
        Nr206 nr206 = new Nr206();
        ListNode head = ListNodeBuilder.buildList(new int[]{1, 2, 3, 4, 5});
        ListNodePrinter.print(nr206.reverseList2(head));
    }

    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode end = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return end;
    }

    public ListNode reverseList2(ListNode head) {
        ListNode prev = null;
        ListNode current = head;
        while (current != null) {
            ListNode tmpNext = current.next;
            current.next = prev;
            prev = current;
            current = tmpNext;
        }

        return prev;
    }
}

File: src/main/problems/ruby/Nr206.rb

# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end
# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  prev = nil
  next_node = head

  while next_node
    next_node, next_node.next, prev = next_node.next, prev, next_node
  end
  prev
end

a = ListNode.new(1)
b = ListNode.new(2)
c = ListNode.new(3)
d = ListNode.new(4)
e = ListNode.new(5)

a.next = b
b.next = c
c.next = d
d.next = e

puts reverse_list(a).val